src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java

```diff
@@ -0,0 +1,195 @@
+            JsonObject responseJson = (JsonObject) Json.decodeValue(responseBody);
+            if (isFailed(responseJson)) {
+                throw new UidCoreClientException(statusCode, "response did not return a successful status");
+            }
+
+            String atoken = getAttestationToken(responseJson);
+            if (atoken == null) {
+                throw new UidCoreClientException(statusCode, "response json does not contain body.attestation_token");
+            }
+            String expiresAt = getAttestationTokenExpiresAt(responseJson);
+            if (expiresAt == null) {
+                throw new UidCoreClientException(statusCode, "response json does not contain body.expiresAt");
+            }
+
+            atoken = new String(decrypt(Base64.getDecoder().decode(atoken), keyPair.getPrivate()), StandardCharsets.UTF_8);
+            LOGGER.info("Attestation successful. Attestation token received.");
+            setAttestationToken(atoken);
+            setAttestationTokenExpiresAt(expiresAt);
+
+            scheduleAttestationExpirationCheck();
+        } catch (AttestationException ae) {
+            throw new UidCoreClientException(ae);
+        } catch (IOException ioe) {
+            throw ioe;
+        } catch (Exception e) {
+            throw new UidCoreClientException(e);
+        }
+    }
+
+    public String getAttestationToken() { return this.attestationToken.get(); }
+    public void setAttestationToken(String atoken) {
+        this.attestationToken.set(atoken);
+    }
+    public void setAttestationTokenExpiresAt(String expiresAt) {
+        this.attestationTokenExpiresAt = Instant.parse(expiresAt);
+    }
+
+    private static String getAttestationToken(JsonObject responseJson) {
+        final JsonObject body = responseJson.getJsonObject("body");
+        if(body == null) return null;
+        return body.getString("attestation_token");
+    }
+
+    private static String getAttestationTokenExpiresAt(JsonObject responseJson) {
+        final JsonObject body = responseJson.getJsonObject("body");
+        if (body == null) return null;
+        return body.getString("expiresAt");
+    }
+
+    private static boolean isFailed(JsonObject responseJson) {
+        return responseJson.getString("status") == null || !responseJson.getString("status").equals("success");
+    }
+
+    private static byte[] decrypt(byte[] payload, PrivateKey privateKey) throws Exception {
+        Cipher cipher = Cipher.getInstance(Const.Name.AsymetricEncryptionCipherClass);
+        cipher.init(Cipher.DECRYPT_MODE, privateKey);
+        return cipher.doFinal(payload);
+    }
+
+    private static KeyPair generateKeyPair() throws NoSuchAlgorithmException {
+        KeyPairGenerator gen = KeyPairGenerator.getInstance(Const.Name.AsymetricEncryptionKeyClass);
+        gen.initialize(2048, new SecureRandom());
+        return gen.generateKeyPair();
+    }
+
+    public void notifyResponseStatusWatcher(int statusCode) {
```

Comment:
I think this whole notification thing should be reworked

- This method should be private
- We should create a new enum called AttestationResponseStatus: Success, InvalidKey, AttestationFailed, Unavailable
- When calling attest internal, we should notify the response watcher with the relevant enum
- We should update the disable watcher to only disable when the enum is InvalidKey or AttestationFailed